src/BalanceTx.purs

```diff
@@ -403,6 +404,21 @@ balanceTxM (UnbalancedTx { transaction: unbalancedTx, utxoIndex }) = do
   feeBuffer :: BigInt
   feeBuffer = fromInt 500000
```

Comment:
unrelated: is this an arbitrary number, what is the logic behind it?

Reply:
We expect that the user has a minimum amount of Ada (this buffer) on top of their transaction, so by the time the prebalancer finishes, should it succeed, there will be some excess Ada (since all NonAda is balanced). In particular, this excess Ada will be at least this buffer.

Hopefully at this point, there will be enough excess Ada to just simply return the Ada to the user, whether by the same utxo (shouldn't cost) or a new utxo (incurring more cost but our buffer should "just" cover it).

I think this buffer can be calculated exactly as an upperbound using this https://github.com/input-output-hk/plutus/blob/8abffd78abd48094cfc72f1ad7b81b61e760c4a0/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs#L723 I should probably make an enhancement issue for this as a use case for our Haskell server.

Reply:
The downside is the input requires more Ada of course.

Reply:
Oh, and yes it is arbitrary at the moment.

Reply:
Maybe you can put a part of this explanation as a comment there?

Reply:
Will do.